### PR TITLE
Phase 6: Tandem Self-Distillation — EMA Teacher for Tandem Samples

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The EMA model is already being maintained throughout training and is the checkpoint used for evaluation. It is a smoother, higher-quality version of the online model. Currently it is used **only at evaluation time**. But since it lives on-device and is updated every step, we can also use it as a **soft teacher** during training.

**Hypothesis:** After EMA begins accumulating (epoch ≥ `ema_start_epoch`), apply a self-distillation KD loss on **tandem samples only**: the EMA model's predictions on tandem nodes serve as soft targets that the online model must also predict. This gives the online model a second, smoother training signal on the hardest OOD split — val_tandem_transfer — at near-zero additional cost (EMA is already computed; we just run a second forward pass on the tandem subset with no_grad).

**Why tandem-only?** Tandem samples are the primary bottleneck (p_tan is our hardest metric). The EMA model has already seen more data per parameter update, is more stable on these samples, and provides a denoised supervision signal. KD on single-foil samples would have lower marginal value since the model already generalizes well there.

**Why this is different from auxiliary losses (dead end #2097):** Deep supervision added auxiliary prediction heads at intermediate layers — a gradient signal that interfered with learning. Self-distillation here is output-level only, after the full forward pass, and uses a soft/smooth target (EMA output) rather than ground truth. The EMA target smooths over noise in the ground truth labels for tandem samples.

**Expected impact:** p_tan −2% to −5%. ~25 LoC. Fully orthogonal to coordinate-based and augmentation changes.

### Step 1: Add config flags

```python
# argparse:
parser.add_argument('--tandem_selfdistill', action='store_true')
parser.add_argument('--tandem_selfdistill_weight', type=float, default=0.1)

# Config dataclass:
tandem_selfdistill: bool = False
tandem_selfdistill_weight: float = 0.1
```

### Step 2: Understand the EMA model in the training loop

Find where the EMA model is defined and updated. Search for:
```bash
grep -n "ema\|EMA\|_base_model\|ema_model" cfd_tandemfoil/train.py | head -40
```

The EMA model is `ema_model` (or similar). It is updated each step after the optimizer step, and used during validation. Note `ema_start_epoch` — EMA only accumulates after this epoch (default 100 in current baseline; may be different in train.py default, check with `grep ema_start_epoch train.py`).

### Step 3: Add the tandem self-distillation loss

In the training loop, AFTER computing the standard loss and AFTER the EMA model update, add:

```python
# Self-distillation KD loss on tandem samples
if cfg.tandem_selfdistill and epoch >= cfg.ema_start_epoch:
    # Identify tandem samples in the batch
    # tandem_mask is already computed earlier in the loop — reuse it
    # tandem_mask: [B] boolean, True for tandem samples
    
    if tandem_mask.any():
        # Get EMA predictions on tandem nodes (no_grad, no gradient through EMA)
        with torch.no_grad():
            ema_out = ema_model(x[tandem_mask], ...)   # same forward args as online model
            # NOTE: ema_model takes the same input signature — copy the forward call
            # from the online model forward pass, substituting ema_model
        
        # Online predictions on tandem nodes (already computed as `out`)
        online_tandem = out[tandem_mask]   # [N_tan, out_dim] — adjust for your tensor shape
        ema_tandem = ema_out               # [N_tan, out_dim]
        
        # KD loss: MSE between online and EMA outputs (on surface nodes only for efficiency)
        # Surface nodes only: use the same surf_mask you apply for surface MAE computation
        kd_loss = F.mse_loss(online_tandem, ema_tandem.detach())
        loss = loss + cfg.tandem_selfdistill_weight * kd_loss
```

**Key implementation detail:** The forward pass signature for `ema_model` is identical to the online model. You do NOT need to recompute anything — just run `ema_model(x[tandem_mask], ...)` with the same arguments used for the online model. Use `torch.no_grad()` to avoid gradients through the EMA model.

**Important:** The KD loss must be added BEFORE `loss.backward()`. If the EMA model update happens after `loss.backward()` in your training loop, compute the EMA forward pass BEFORE the optimizer step but AFTER the online forward pass. Check the exact order in the training loop with:
```bash
grep -n "ema\|backward\|optimizer.step\|loss" cfd_tandemfoil/train.py | grep -A2 -B2 "backward"
```

**Memory:** The EMA model forward pass on tandem-only samples (typically 30-50% of batch) costs ~1/3 extra compute. If you see OOM, filter to surface nodes only before computing KD loss, or reduce batch size by 10%.

### Step 4: Sweep over distillation weight

Run 3 seeds: w=0.05, w=0.10, w=0.20:

```bash
# w=0.10, seed 42 (primary)
cd cfd_tandemfoil && python train.py --agent edward \
  --wandb_name "edward/tandem-selfdistill-w0.10-s42" \
  --wandb_group phase6/tandem-selfdistill --seed 42 \
  --tandem_selfdistill --tandem_selfdistill_weight 0.10 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aft_foil_srf_context

# w=0.05, seed 42: same but --tandem_selfdistill_weight 0.05
# w=0.20, seed 42: same but --tandem_selfdistill_weight 0.20

# Control seed 42: same WITHOUT --tandem_selfdistill
```

**W&B group:** `phase6/tandem-selfdistill`

Report p_in, p_oodc, p_tan, p_re + W&B run IDs for all 4 runs. If w=0.10 seed 42 shows clear improvement, run a second seed (73) on the best weight for 2-seed averaging.

**Watch for:** If p_tan degrades monotonically as w increases, the KD loss is over-constraining the online model. If p_oodc degrades with no p_tan improvement, the KD loss is hurting generalization elsewhere — try restricting to surface pressure nodes only (not velocity/pressure volume nodes).

## Baseline

Current single-model baseline (PR #2127, AftSRF KNN context K=8, 2-seed mean):

| Metric | 2-seed mean | Target |
|--------|-------------|--------|
| p_in | **13.02** | < 13.02 |
| p_oodc | **7.62** | < 7.62 |
| **p_tan** | **29.91** | **< 29.91** ← primary |
| p_re | **6.47** | < 6.47 |

W&B: `zosxwjmm` (s42), `twilqf1x` (s73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent edward --seed 42 \
  --wandb_name "edward/baseline-aft-srf-ctx" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aft_foil_srf_context
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results: Tandem Self-Distillation (EMA Teacher)

### Setup
- 4 runs: 3 KD weights (w=0.05, 0.10, 0.20) + 1 control, all seed 42
- All runs: `--ema_start_epoch 100 --ema_decay 0.999 --cosine_T_max 160 --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aft_foil_srf_context --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 --surface_refine_hidden 192 --surface_refine_layers 3 --residual_prediction`
- KD runs add: `--tandem_selfdistill --tandem_selfdistill_weight <w>`
- Implementation: After EMA activation, MSE between online and EMA backbone predictions on surface pressure (channel 2) for tandem samples only

### Critical Issue: GPU Speed Variance → Post-Cosine Degradation

GPUs 1-3 ran at ~43-45s/epoch (ctrl), reaching **220-252 epochs** before the 180-min timeout. GPU 0 (w=0.05) ran at ~80s/epoch, reaching only **136 epochs**. The cosine LR schedule ends at epoch 160 (`cosine_T_max=160`).

**Runs that went far beyond epoch 160 showed severe degradation**, especially on OOD metrics. This affects ALL runs on GPUs 1-3, including the control. The model overfits to in-distribution data after the LR schedule bottoms out, and the EMA tracks this degradation.

### Results — Surface MAE (Pressure)

| Run | Epochs | p_in | p_oodc | p_tan | p_re | val/loss | W&B Run ID |
|-----|--------|------|--------|-------|------|----------|------------|
| **w=0.05 s42** | **136** | **13.6** | **8.3** | **30.4** | **6.7** | **0.396** | `swsmtceu` |
| w=0.10 s42 | 219 | 14.1 | 10.8 | 38.2 | 25.9 | 0.735 | `ewrhonfl` |
| w=0.20 s42 | 220 | 14.9 | 10.4 | 35.6 | 25.6 | 0.711 | `d6x7l5cg` |
| ctrl s42 | 252 | 15.0 | 9.9 | 36.5 | 25.4 | 0.713 | `ivs2xoxz` |
| **Baseline (2-seed)** | — | **13.02** | **7.62** | **29.91** | **6.47** | — | — |

⚠️ **Only the w=0.05 run is a valid result** (136 epochs ≈ within cosine schedule). The other 3 runs suffered post-cosine degradation and their metrics are not meaningful.

### Analysis of the Valid Run (w=0.05)

The w=0.05 run does not beat any baseline metric:
- p_in: 13.6 vs 13.02 (+4.5%)
- p_oodc: 8.3 vs 7.62 (+8.9%)
- p_tan: 30.4 vs 29.91 (+1.6%)
- p_re: 6.7 vs 6.47 (+3.6%)

However, this is a single run on a slow GPU that only had **36 epochs of KD** (EMA started at epoch 100, training stopped at epoch 136). The KD signal barely had time to influence training. A fair test would need the same epoch count as the baseline (~150-170 epochs).

### Why GPUs 1-3 Were So Fast

With `--aft_foil_srf_context`, the per-epoch time dropped from the expected ~63s to ~43-45s. The KD forward pass added ~17s overhead on KD runs (62-63s/epoch), but even KD runs went to 219-220 epochs — well past the cosine schedule endpoint.

### Root Cause: cosine_T_max vs Actual Epochs

The core problem: `cosine_T_max=160` assumes ~170 epochs in 180 min (~63s/epoch). With `--aft_foil_srf_context`, epochs are faster (~45s/epoch), reaching 250+ epochs. Training 90+ epochs beyond the cosine schedule causes severe OOD degradation.

### Suggested Follow-ups

1. **Re-run with `--cosine_T_max 240`** to match the actual epoch count for these GPUs, ensuring the LR schedule covers the full training duration
2. **OR stop training at epoch 160** by capping max epochs (requires code change or reduced `SENPAI_TIMEOUT_MINUTES`)
3. **The KD idea itself is inconclusive** — the one valid run didn't have enough KD epochs to test the hypothesis. If re-run with proper LR schedule, results may differ
4. **The GPU speed variance issue is general** — any experiment using `--aft_foil_srf_context` will face this problem unless cosine_T_max is adjusted

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2135 because ADVISOR: Closing — inconclusive due to post-cosine degradation. Only w=0.05 (136 epochs) was valid; it regressed on all metrics (p_tan +1.6%, p_oodc +8.9%). Also confounded by the aft_foil_srf_context guard bug (frieren #2134) — the context head was never applied, so epoch timing differed from expected. The KD hypothesis had too little exposure time (36 epochs) but the initial signal is negative. Good catch on the post-cosine degradation issue — this is real and affects all fast-GPU runs with cosine_T_max=160. Reassigning to a fresh experiment..
